Increase robustness of coordinate transformation class use

lindsayad opened the pull request:
Closes #21739 #21975

GiudGiud committed:
Add test for coordinate transform of user objects transfers, refs #

lindsayad committed:
Do not set the coordinate system in prepare if it's already been set

This allows SAM to continue programmatically setting their mesh
information

Refs changes in #20451

lindsayad committed:
Warn user about using non-translation transforms without MooseCoordTransform

Also call `getAppInfo` in `initialSetup` and not in `execute`

Closes #21739

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1132088/) on ce6fe61 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 004f3bd8409919b662f8e5020b65aa6c4bd9f40b`

lindsayad committed:
Notify associated transfers when multiapp reset

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1132369/) on 0e1720b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21788/site)

This comment will be updated on new commits.

lindsayad commented:
@GiudGiud I know you self-assigned yourself on #21739 but fixing up the coordinate transform setting for SAM kind of necessitated addressing #21739. I think you said you had done some work on this but I didn't see any commits referencing #21739. If you have already done some work, we can look at what each other has done and see about combining the best of our work

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -282,10 +293,28 @@ MultiAppTransfer::getAppInfo()
+    auto check_transform_compatibility = [this](const MooseCoordTransform & transform)
+    {
+      if (transform.hasNonTranslationTransformation() && !usesMooseCoordTransform())
+        mooseDoOnce(
```
I would remove this mooseDoOnce
Since this is shared, it will only output once for the whole simulation, instead of once per transfer type as we had previously

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -282,10 +293,28 @@ MultiAppTransfer::getAppInfo()
+    {
+      if (transform.hasNonTranslationTransformation() && !usesMooseCoordTransform())
+        mooseDoOnce(
+            mooseWarning("'",
```
```suggestion
            mooseWarning("Transfer '",
```

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -282,10 +293,28 @@ MultiAppTransfer::getAppInfo()
+                         name(),
+                         "' of type '",
+                         type(),
+                         "' has non-translation transformations but it does not leverage the "
+                         "'MooseCoordTransform' class. Your data transfers may be inaccurate"));
```
```suggestion
                         "' has non-translation transformations but it does not implement coordinate transformations using the 'MooseCoordTransform' class. Your data transfers will not be performed in the expected transformed frame"));
```

GiudGiud commented on framework/src/transfers/MultiAppVariableValueSampleTransfer.C:
```diff
@@ -52,6 +52,8 @@ MultiAppVariableValueSampleTransfer::MultiAppVariableValueSampleTransfer(
 void
 MultiAppVariableValueSampleTransfer::initialSetup()
 {
+  MultiAppTransfer::initialSetup();
```
is there a way to programmatically force all our users' derived classes to call the base class routine first? like some sort of 'override' keyword but like 'override_but_call_anyway'

lindsayad replied:
No there is not

GiudGiud replied:
another let down from C++ lolol

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -521,6 +522,9 @@ MultiApp::preTransfer(Real /*dt*/, Real target_time)
     for (auto & app : _reset_apps)
       resetApp(app);
+
+    for (auto * const transfer : _associated_transfers)
```
because I removed `getAppInfo` from all the `execute` methods and moved it to `initialSetup`. That change if fine unless an app gets reset in which case you get to restore the app info

lindsayad committed:
Apply suggestions from code review

- Remove do-once from transfer coord transform warning
- reword transfer warning

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1136964/) on 8614e10 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format eb0128f5d3aec72cce02c9a92e2c2b838d629325`

lindsayad commented on framework/doc/content/syntax/Transfers/index.md:
```diff
@@ -142,7 +142,6 @@ processes described here are:
 - [MultiAppNearestNodeTransfer.md]
 - [MultiAppPostprocessorInterpolationTransfer.md]
 - [MultiAppProjectionTransfer.md]
-- [MultiAppUserObjectTransfer.md]
```
what test case? I'm surprised that the results were wrong. I would have hoped that your results would either be correct (compatible coordinate systems) or you would have gotten an error in `mapBack`

lindsayad replied:
The latter was the reason I reverted `MooseCoordTransfrom` support of `MultiAppUserObjectTransfer` ... although now I'm thinking of reverting the revert

GiudGiud replied:
results were incorrect, no error in mapBack. I can contribute my test tomorrow

I m in favor of coord transforms for UO transfers :)

lindsayad replied:
Yea it would be great to get that test

lindsayad committed:
When applying a common parameter, copy its value of _set_by_add_param

lindsayad committed:
Create MultiCoordTransform

lindsayad committed:
Allow skipping of coordinate collapsing

lindsayad committed:
Ensure that we are applying transforms for UO from-multiapp

lindsayad committed:
Allow MultiApp::getBoundingBox to take a coord transform arg

lindsayad committed:
Add localAppNumber API

This is very useful for making sure we don't index our transforms
vector out of bounds in derived classes of MultiAppTransfer

lindsayad committed:
Remove extraneous loop over non-existent app number

An index at numGlobalApps() will never have a corresponding app
because we are doing C++ not Fortran

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1140510/) on 34def54 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>925a17</th>
      <th colspan=4>#21788 34def5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.98%</td>
      <td>83.98%</td>
      <td>+0.00%</td>
      <td>92.92%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>78631</td>
      <td>78705</td>
      <td>+74</td>
      <td>210</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14998</td>
      <td>15010</td>
      <td>+12</td>
      <td>16</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21788/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21788/coverage/framework)


## Modules coverage

### Thermal hydraulics

<table>
  <thead>
    <tr>
      <th></th>
      <th>925a17</th>
      <th colspan=4>#21788 34def5</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>91.10%</td>
      <td>91.10%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12224</td>
      <td>12229</td>
      <td>+5</td>
      <td>5</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1194</td>
      <td>1194</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21788/coverage_diff/thermal_hydraulics)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21788/coverage/thermal_hydraulics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21788/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad committed:
Allow skipping coordinate collapsing in more transfers

Supports some of our application tests

lindsayad commented:
Ok I have patches ready for
- [x] subchannel
- [x] pronghorn
- [x] sabertooth

GiudGiud commented:
I guess I owe an update here.
I m delayed on the review, we have some publications and reports to take care off until mid-next week, and this is too big to take care mid-procastinating-writing. Sorry for everyone this is impacting.

lingzou commented:
> I guess I owe an update here.
> I m delayed on the review, we have some publications and reports to take care off until mid-next week, and this is too big to take care mid-procastinating-writing. Sorry for everyone this is impacting.

@GiudGiud Thanks for the update. Understood, and please take your time.

GiudGiud commented on framework/src/utils/MooseCoordTransform.C:
```diff
@@ -399,43 +409,49 @@ MooseCoordTransform::operator()(const Point & point) const
-  _destination_coord_type = destination_coord_transform._coord_type;
-  _destination_r_axis = destination_coord_transform._r_axis;
-  _destination_z_axis = destination_coord_transform._z_axis;
+  _destination_coord_type = destination_coord_transform._single_app_transform._coord_type;
```
this makes me feel that a Multi(App)CoordTransform should have two attributes:
- origin single-app-transform
- desitrination single-app-transform
and work with those

lindsayad replied:
:+1:

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -64,6 +63,16 @@ MultiAppTransfer::addBBoxFactorParam(InputParameters & params)
+{
+  params.addParam<bool>("skip_coordinate_collapsing",
+                        false,
+                        "Whether to skip coordinate collapsing when performing mapping and inverse "
```
```suggestion
                        "Whether to skip coordinate collapsing (translation and rotation are still performed, only XYZ, RZ etc collapsing is skipped) when performing mapping and inverse "
```

GiudGiud reviewed (changes requested):
Current design for the MultiCoordTransform (to be renamed MultiAppCoordTransform)
composition with an MooseCoordTransform for the origin and additional attributes for the destination. 

for the Multi(App)CoordTransform I could see two alternate designs:
- a class inheriting the MooseCoordTransform for the origin and adding the attributes from the destination
this would avoid the friend class declaration but would not achieve much more
- keep the composition design here but store origin and destination transform as attributes (and keep the friend class for easy access)
IMO either will adapt easier to between_multiapps transfers. 
I m a bigger fan of the second one

@loganharbour what do you think?

loganharbour reviewed (dismissed):
Add check comparing the coord system in the problem and the mesh, erroring if inconsistent

lindsayad committed:
MooseCoordTransform -> MooseAppCoordTransform

lindsayad committed:
MultiCoordTransform -> MultiAppCoordTransform

lindsayad committed:
Make the destination coord transform an attribute per GiudGiud suggestion

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1158223/) on 13469fd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83889eae5385918c147a11cb87bef4b3896383b8`

lindsayad commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -64,6 +63,16 @@ MultiAppTransfer::addBBoxFactorParam(InputParameters & params)
+void
+MultiAppTransfer::addSkipCoordCollapsingParam(InputParameters & params)
+{
+  params.addParam<bool>("skip_coordinate_collapsing",
+                        false,
+                        "Whether to skip coordinate collapsing when performing mapping and inverse "
+                        "mapping coordinate transformation operations. This parameter should only "
+                        "be set by users who really know what they're doing.");
```
so you think we should have two parameters: `skip_main_app_coordinate_collapsing` and `skip_sub_app_coordinate_collapsing`? How would that work for a between-multiapp? Just have that one use the value of `skip_sub_app_coordinate_collapsing`?

GiudGiud replied:
not suuuuure
skip_origin_coordinate_collapsing and skip_target_coordinate_collapsing then?

coordinate transforms is going to be the most siblings-transfer-enabled feature haha

lindsayad replied:
with those names how would a user know which boolean was being applied to which sibling app?

GiudGiud replied:
so the siblings are from_multiapp and to_multiapp so the from_multiapp is naturally the origin coordinate, and the to_ the target coordinate

(except you can have bi-directional transfers still :O  but those are rare)

lindsayad replied:
What about `skip_from_app_coordinate_collapsing` an `skip_to_app_coordinate_collapsing`?

GiudGiud replied:
works for me

lindsayad replied:
Thinking more about this ... I don't think having the two distinct options makes sense. If collapsing is happening, it means that one app will collapse its coordinates into the other's. So in truth, skipping coordinate collapsing only applies to one of the paired apps; the other app experiences no collapsing ... so a single `skip_coordinate_collapsing` option makes sense to me.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1158291/) on 6023ea1 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21788/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 83889eae5385918c147a11cb87bef4b3896383b8`

lindsayad committed:
Output warning once per multiapp and clarify collapsing param string

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad committed:
setDestinationCoordinateSystem -> setDestinationCoordTransform

lindsayad committed:
Add error to catch different param settings in Mesh and Problem

Closes #21975

lindsayad committed:
Ensure destination setting works on all processes

The problem addressed in this commit was revealed in a phase
field split test, when in a two process run, process 1 did not have
any sub-apps. We should be able to construct `MooseAppCoordTransforms`
representing the non-translation transformations for both from and
to applications on all processes by broadcasting from rank 0 which,
I believe, should always have a sub-app. The changes in this commit
should help enable additition of coordinate transformation capabilities
to other transfer classes which query transformation information
for all global sub-apps

GiudGiud commented on framework/include/utils/MooseAppCoordTransform.h:
```diff
@@ -15,33 +15,74 @@
+                     short int,
+                     std::array<Real, 3>,
+                     int,
+                     unsigned int,
```
dont want to use the enum defined just above?

GiudGiud replied:
alright works for me

GiudGiud commented on framework/src/transfers/MultiAppInterpolationTransfer.C:
```diff
@@ -16,7 +16,7 @@
-#include "MooseCoordTransform.h"
+#include "MooseAppCoordTransform.h"
 
 #include "libmesh/parallel_algebra.h"
```
should you add the skipCoordTransform parameters everywhere?
  MultiAppTransfer::addSkipCoordCollapsingParam(params);

lindsayad replied:
I modeled this design after the bounding box expansion factor, which you were the conceptual pioneer of. I personally like only adding the parameter to classes for which both of the following are true:

1. The class supports the coordinate transformation object
2. Coordinate collapsing is potentially "problematic", e.g. `mapBack` is called, and/or users have leveraged a lack of coordinate collapsing in the past when desiging their inputs

I think that minimizing the use of `skip_coordinate_collapsing` is a good thing ... or else coordinate collapsing within the transformation object is just a bad feature if we find ourselves trying to avoid it a lot. And maybe it is 🤷 I've wondered that to myself.

GiudGiud replied:
kk works for me.
transfers are very often interchangeable (as long as you dont care about projection errors), so if 2 applied to 2 transfers, it likely could apply to all of them

GiudGiud commented on framework/src/transfers/MultiAppNearestNodeTransfer.C:
```diff
@@ -15,7 +15,7 @@
 #include "MooseVariableFE.h"
-#include "MooseCoordTransform.h"
+#include "MooseAppCoordTransform.h"
 
```
same question

GiudGiud commented on framework/src/transfers/MultiAppProjectionTransfer.C:
```diff
@@ -15,7 +15,7 @@
 #include "SystemBase.h"
-#include "MooseCoordTransform.h"
+#include "MooseAppCoordTransform.h"
 
```
same

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -64,6 +63,18 @@ MultiAppTransfer::addBBoxFactorParam(InputParameters & params)
+      "XYZ, RZ etc collapsing is skipped) when performing mapping and inverse "
+      "mapping coordinate transformation operations. This parameter should only "
+      "be set by users who really know what they're doing.");
+}
```
add it to advanced group?

GiudGiud commented on framework/src/utils/MooseAppCoordTransform.C:
```diff
@@ -344,107 +377,128 @@ MooseCoordTransform::computeRS()
+        _destination_app_transform->_coord_type == Moose::COORD_RSPHERICAL)) ||
+      (_our_app_transform._coord_type == Moose::COORD_RZ &&
+       _destination_app_transform->_coord_type == Moose::COORD_RSPHERICAL))
     mooseError("Coordinate collapsing occurred in going to the reference space. There is no unique "
```
a pronghorn test fails this, will need to skip the coordinate collapse there

GiudGiud commented on framework/src/utils/MooseAppCoordTransform.C:
```diff
@@ -115,16 +117,17 @@ MooseCoordTransform::setRotation(const Real alpha, const Real beta, const Real g
 
   if (must_rotate_axes && !axes_rotated)
-    mooseError("Unsupported manual angle prescription in 'MooseCoordTransform::setRotation'");
+    mooseError("Unsupported manual angle prescription in 'MooseAppCoordTransform::setRotation'");
```
```suggestion
    mooseError("Unsupported manual angle prescription in 'MooseAppCoordTransform::setRotation.' "
                        "Supported rotation is (alpha, beta, gamma) = (0, 90, 0)");
```

GiudGiud replied:
well I wasnt adding just a point there?

lindsayad replied:
Ah indeed not

lindsayad replied:
Now I understand why that space was there!

GiudGiud reviewed (commented):
no major comments, this is ready as soon as app patches are ready. 
With allow_warnings we should be able to merge patches and get the apps green on this PR

GiudGiud commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -271,10 +299,28 @@ MultiAppTransfer::getAppInfo()
     const auto & ex_to_transform = *_to_transforms[0];
 
-    for (auto * const from_transform : _from_transforms)
+    auto check_transform_compatibility = [this](const MultiCoordTransform & transform)
```
your solution is better

GiudGiud commented:
@loganharbour your review has been addressed

lindsayad committed:
Add coord collapsing to advanced group and add info to rotation error

GiudGiud commented:
Cant you pre-position skip_coordinate_collapsing as an unused parameter with allow_unused = true? It should go all the way to master in the particular app that way?

lindsayad commented:
can give that a whirl

GiudGiud reviewed (approved):
double approval

lindsayad commented:
thanks for the reviews!
